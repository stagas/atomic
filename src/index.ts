interface Task {
  context: any
  resolve: (value: any) => void
  reject: (reason?: any) => void
  args: any[]
}

/**
 * Makes a function atomic.
 *
 * @param signalClosure A function that receives the signal object
 *  from an AbortController and returns the function to become atomic.
 * @param maxTimeMs Time in milliseconds to timeout the operation. Will also signal abort.
 * @returns An atomic function
 */
export const atomic = (
  signalClosure: (signal: AbortSignal) => (...args: any[]) => Promise<any>,
  maxTimeMs?: number
) => {
  const queue: Task[] = []

  let flushing = false

  const flush = async () => {
    if (flushing) return
    flushing = true

    const task = queue.shift()!
    const abortController = new AbortController()

    try {
      const fn = signalClosure(abortController.signal)
      let promise = fn.apply(task.context, task.args)
      if (maxTimeMs != null) {
        promise = Promise.race([
          promise,
          new Promise((_, reject) => setTimeout(reject, maxTimeMs, new Error('Timed out'))),
        ])
      }
      task.resolve(await promise)
    } catch (error) {
      abortController.abort()
      task.reject(error)
    } finally {
      flushing = false
      if (queue.length) flush()
    }
  }

  return function (this: any, ...args: any[]): Promise<any> {
    let resolve!: Task['resolve'], reject!: Task['reject']
    const promise = new Promise((_resolve, _reject) => {
      resolve = _resolve
      reject = _reject
    })
    queue.push({ context: this, resolve, reject, args })
    flush()
    return promise
  }
}
