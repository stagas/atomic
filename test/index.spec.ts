import { atomic } from '../src'

describe('atomic(fn)', () => {
  it('makes the function atomic', async () => {
    const results = []
    let i = 0
    const fn = atomic(() => async () => {
      const a = i
      await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
      results.push(a + 1)
      i = a + 1
    })
    fn()
    fn()
    fn()
    await fn()
    expect(i).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('receives parameters', async () => {
    const results = []
    let i = 0
    const fn = atomic(() => async (x: number) => {
      const a = i
      await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
      results.push(a + 1 + x)
      i = a + 1
    })
    fn(1)
    fn(1)
    fn(1)
    await fn(1)
    expect(i).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('alternate', async () => {
    const results = []
    let i = 0
    const fn = atomic(() => async () => {
      const a = i
      await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
      results.push(a + 1)
      i = a + 1
    })
    const promises = [fn(), fn(), fn(), fn()]
    await Promise.all(promises)
    expect(i).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('returns result', async () => {
    const results = []
    let i = 0
    const fn = atomic(() => async () => {
      const a = i
      await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
      results.push(a + 1)
      i = a + 1
      return i
    })
    fn()
    fn()
    fn()
    const result = await fn()
    expect(i).toEqual(4)
    expect(result).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('throwing doesnt destroy queue', async () => {
    const results = []
    let i = 0
    let times = 0
    const fn = atomic(() => async () => {
      const a = i
      times++
      if (a === 2) throw new Error('problem')
      await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
      results.push(a + 1)
      i = a + 1
      return i
    })
    await Promise.allSettled([fn(), fn(), fn(), fn()])
    expect(i).toEqual(2)
    expect(times).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('uses this', async () => {
    const results = []
    const obj = {
      i: 0,
      fn: atomic(
        () =>
          async function () {
            const a = this.i
            await new Promise(resolve => setTimeout(resolve, Math.random() * 100))
            results.push(a + 1)
            this.i = a + 1
            return a
          }
      ),
    }
    await Promise.allSettled([obj.fn(), obj.fn(), obj.fn(), obj.fn()])
    expect(obj.i).toEqual(4)
    expect(results).toMatchSnapshot()
  })

  it('times out after maxTimeMs', async () => {
    const results = []
    let i = 0
    let times = 0
    const fn = atomic(
      signal => async () => {
        const a = i
        times++
        await new Promise(resolve => setTimeout(resolve, a === 2 ? 300 : 100))
        if (signal.aborted) return
        results.push(a + 1)
        i = a + 1
        return i
      },
      200
    )
    const output = await Promise.allSettled([fn(), fn(), fn(), fn()])
    await new Promise(resolve => setTimeout(resolve, 200))
    expect(i).toEqual(2)
    expect(times).toEqual(4)
    expect(results).toMatchSnapshot()
    expect(output).toMatchSnapshot()
  })
})
