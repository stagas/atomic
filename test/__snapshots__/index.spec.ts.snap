// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atomic(fn) alternate 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`atomic(fn) makes the function atomic 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`atomic(fn) receives parameters 1`] = `
Array [
  2,
  3,
  4,
  5,
]
`;

exports[`atomic(fn) returns result 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`atomic(fn) throwing doesnt destroy queue 1`] = `
Array [
  1,
  2,
]
`;

exports[`atomic(fn) times out after maxTimeMs 1`] = `
Array [
  1,
  2,
]
`;

exports[`atomic(fn) times out after maxTimeMs 2`] = `
Array [
  Object {
    "status": "fulfilled",
    "value": 1,
  },
  Object {
    "status": "fulfilled",
    "value": 2,
  },
  Object {
    "reason": [Error: Timed out],
    "status": "rejected",
  },
  Object {
    "reason": [Error: Timed out],
    "status": "rejected",
  },
]
`;

exports[`atomic(fn) uses this 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;
